File: csharp/CsharpCommandEngine/UsageCommand.cs
using System;
using System.Collections.Generic;
using CsharpFileCommands;
using Validators;

namespace CsharpCommandEngine
{
    public class UsageCommand : AbstractCommand
    {
        private static readonly string ArgumentLimitMessage = "The command takes strictly less than 2 arguments.";

        public new static readonly string UsageString =
                   "usage command displays available commands or the way to use the command in its argument.\n" +
                   "Use it the following way:\n" +
                   "\t'usage' - displays the list of available commands.\n" +
                   "\t'usage <command>' - displays the description and way to use for <command>.\n" +
                   "\tNOTE: " + ArgumentLimitMessage;

        protected override void SetupValidation()
        {
            Validation.AddValidator(new ArgumentLimitValidator(Arguments, 2, ArgumentLimitMessage));
        }

        public UsageCommand(List<string> args) : base(args)
        {
            SetupValidation();
        }

        public override void Execute()
        {
            if (!Validation.Validate())
                return;

            if (Arguments.Count == 0)
                Console.WriteLine("Available commands: {0}.", GetAvailableCommands());
            else
                switch (Arguments[0])
                {
                    case "usage":
                        Console.WriteLine(UsageString);
                        break;
                    case "exit":
                        Console.WriteLine(ExitCommand.UsageString);
                        break;
                    case "copy":
                        Console.WriteLine(CopyCommand.UsageString);
                        break;
                    case "move":
                        Console.WriteLine(MoveCommand.UsageString);
                        break;
                    case "rename":
                        Console.WriteLine(RenameCommand.UsageString);
                        break;
                    case "delete":
                        Console.WriteLine(DeleteCommand.UsageString);
                        break;
                    case "list":
                        Console.WriteLine(ListDirectoryCommand.UsageString);
                        break;
                    default:
                        Console.WriteLine("Usage doesn't know this command.");
                        Console.WriteLine("Available commands: {0}.", GetAvailableCommands());
                        break;
                }
        }
    }
}

File: python/PythonCommandEngine/UsageCommand.py
from .ExitCommand import *
from Validators.ArgumentLimitValidator import *
from PythonFileCommands.DeleteCommand import *
from PythonSearchCommands.ListCommand import *
from PythonSearchCommands.LookupCommand import *
from PythonFileCommands.CopyCommand import *


class UsageCommand(AbstractCommand):
    def __init__(self, args):
        super().__init__(args)
        self._usage_validation()

    @classmethod
    def _argument_limit_message(cls):
        return "The command takes strictly less than 2 arguments."

    @classmethod
    def usage_string(cls):
        return "usage command displays available commands or the way to use the command in its argument.\n" + \
               "Use it the following way:\n" + \
               "\t'usage' - displays the list of available commands.\n" + \
               "\t'usage <command>' - displays the description and way to use for <command>.\n" + \
               "\t" + cls._note_preface() + cls._argument_limit_message()

    @staticmethod
    def _available_commands():
        return "Available commands: " + ', '.join(["usage", "exit", "delete", "list", "lookup", "copy"]) + "."

    @staticmethod
    def _command_unknown_preface():
        return "Usage doesn't know this command."

    def _usage_validation(self):
        self._validation.add_validator(ArgumentLimitValidator(self.arguments, 2, self._argument_limit_message()))

    @staticmethod
    def _choose_usage(choice):
        return {
            "usage": UsageCommand.usage_string(),
            "exit": ExitCommand.usage_string(),
            "delete": DeleteCommand.usage_string(),
            "list": ListCommand.usage_string(),
            "lookup": LookupCommand.usage_string(),
            "copy": CopyCommand.usage_string(),
            "": UsageCommand._available_commands()
        }.get(choice, UsageCommand._command_unknown_preface() + "\n\t" + UsageCommand._available_commands())

    def execute(self):
        if not self._validation.validate():
            return
        print(self._choose_usage(self.arguments[0] if self.arguments.__len__() > 0 else ""))


